### ISSUE TITLE
Latest version does not run on iOS13: -[MTLSimDevice counterSets]: unrecognized selector sent to instance

### ISSUE THREAD

**Issue by flolag** [KhronosGroup/MoltenVK#1518]
Hello,

I am currently integrating MoltenVK in an existing GLES2 SDK. The SDK we are building has to support iOS11 as a minimum.

A recent commit (6ae1745a9ceaf5b5ea52777d95074154118f7919 - 21 Jul 2021) introduced a call to `-[MTLSimDevice counterSets]` which is not supported before iOS14.

On iOS13 (on simulator) this causes the call to vkCreateInstance to crash with the message:
`-[MTLSimDevice counterSets]: unrecognized selector sent to instance`

Full stack trace:

```
2022-02-04 10:19:42.628184+0100 Cube[5219:71769] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[MTLSimDevice counterSets]: unrecognized selector sent to instance 0x7fdb1f104560'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff23e3de6e __exceptionPreprocess + 350
	1   libobjc.A.dylib                     0x00007fff512a19b2 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff23e5eb94 -[NSObject(NSObject) doesNotRecognizeSelector:] + 132
	3   CoreFoundation                      0x00007fff23e4286c ___forwarding___ + 1436
	4   CoreFoundation                      0x00007fff23e44b58 _CF_forwarding_prep_0 + 120
	5   Cube                                0x000000010c6bd974 _ZN17MVKPhysicalDevice15initCounterSetsEv + 116
	6   Cube                                0x000000010c6bb664 _ZN17MVKPhysicalDeviceC2EP11MVKInstancePU19objcproto9MTLDevice11objc_object + 292
	7   Cube                                0x000000010c6bef05 _ZN17MVKPhysicalDeviceC1EP11MVKInstancePU19objcproto9MTLDevice11objc_object + 37
	8   Cube                                0x000000010c614dec _ZN11MVKInstanceC2EPK20VkInstanceCreateInfo + 1372
	9   Cube                                0x000000010c622ced _ZN11MVKInstanceC1EPK20VkInstanceCreateInfo + 29
	10  Cube                                0x000000010c5fd0eb vkCreateInstance + 75
	11  Cube                                0x000000010c5c5291 demo_init_vk + 1793
	12  Cube                                0x000000010c5c3b8d demo_init + 1581
	13  Cube                                0x000000010c5c26e7 demo_main + 39
	14  Cube                                0x000000010c5c25d3 -[DemoViewController viewDidLoad] + 243
	15  UIKitCore                           0x00007fff48cc294e -[UIViewController _sendViewDidLoadWithAppearanceProxyObjectTaggingEnabled] + 83
	16  UIKitCore                           0x00007fff48cc786c -[UIViewController loadViewIfRequired] + 1084
	17  UIKitCore                           0x00007fff48cc7c89 -[UIViewController view] + 27
	18  UIKitCore                           0x00007fff493ab2d5 -[UIWindow addRootViewControllerViewIfPossible] + 326
	19  UIKitCore                           0x00007fff493aa8fe -[UIWindow _updateLayerOrderingAndSetLayerHidden:actionBlock:] + 219
	20  UIKitCore                           0x00007fff493ab989 -[UIWindow _setHidden:forced:] + 362
	21  UIKitCore                           0x00007fff493bedc5 -[UIWindow _mainQueue_makeKeyAndVisible] + 42
	22  UIKitCore                           0x00007fff4936b08f -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] + 4209
	23  UIKitCore                           0x00007fff49370b71 -[UIApplication _runWithMainScene:transitionContext:completion:] + 1226
	24  UIKitCore                           0x00007fff48a7912d -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] + 122
	25  UIKitCore                           0x00007fff48f84b61 _UIScenePerformActionsWithLifecycleActionMask + 83
	26  UIKitCore                           0x00007fff48a79c3f __101-[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:]_block_invoke + 198
	27  UIKitCore                           0x00007fff48a7964e -[_UISceneLifecycleMultiplexer _performBlock:withApplicationOfDeactivationReasons:fromReasons:] + 296
	28  UIKitCore                           0x00007fff48a79a6c -[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:] + 818
	29  UIKitCore                           0x00007fff48a79301 -[_UISceneLifecycleMultiplexer uiScene:transitionedFromState:withTransitionContext:] + 345
	30  UIKitCore                           0x00007fff48a7d8f2 __186-[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:]_block_invoke_2 + 178
	31  UIKitCore                           0x00007fff48e9ac4d +[BSAnimationSettings(UIKit) tryAnimatingWithSettings:actions:completion:] + 852
	32  UIKitCore                           0x00007fff48fa32be _UISceneSettingsDiffActionPerformChangesWithTransitionContext + 240
	33  UIKitCore                           0x00007fff48a7d60d __186-[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:]_block_invoke + 153
	34  UIKitCore                           0x00007fff48fa31c1 _UISceneSettingsDiffActionPerformActionsWithDelayForTransitionContext + 84
	35  UIKitCore                           0x00007fff48a7d47b -[_UIWindowSceneFBSSceneTransitionContextDrivenLifecycleSettingsDiffAction _performActionsForUIScene:withUpdatedFBSScene:settingsDiff:fromSettings:transitionContext:lifecycleActionType:] + 381
	36  UIKitCore                           0x00007fff488d1e78 __64-[UIScene scene:didUpdateWithDiff:transitionContext:completion:]_block_invoke + 657
	37  UIKitCore                           0x00007fff488d0a37 -[UIScene _emitSceneSettingsUpdateResponseForCompletion:afterSceneUpdateWork:] + 253
	38  UIKitCore                           0x00007fff488d1ba2 -[UIScene scene:didUpdateWithDiff:transitionContext:completion:] + 210
	39  UIKitCore                           0x00007fff4936f0b9 -[UIApplication workspace:didCreateScene:withTransitionContext:completion:] + 512
	40  UIKitCore                           0x00007fff48ec177c -[UIApplicationSceneClientAgent scene:didInitializeWithEvent:completion:] + 361
	41  FrontBoardServices                  0x00007fff36d03d2e -[FBSSceneImpl _callOutQueue_agent_didCreateWithTransitionContext:completion:] + 419
	42  FrontBoardServices                  0x00007fff36d29dc1 __86-[FBSWorkspaceScenesClient sceneID:createWithParameters:transitionContext:completion:]_block_invoke.154 + 102
	43  FrontBoardServices                  0x00007fff36d0e757 -[FBSWorkspace _calloutQueue_executeCalloutFromSource:withBlock:] + 220
	44  FrontBoardServices                  0x00007fff36d29a52 __86-[FBSWorkspaceScenesClient sceneID:createWithParameters:transitionContext:completion:]_block_invoke + 355
	45  libdispatch.dylib                   0x00007fff520dc8cb _dispatch_client_callout + 8
	46  libdispatch.dylib                   0x00007fff520df292 _dispatch_block_invoke_direct + 285
	47  FrontBoardServices                  0x00007fff36d4f6e9 __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 30
	48  FrontBoardServices                  0x00007fff36d4f3d7 -[FBSSerialQueue _queue_performNextIfPossible] + 441
	49  FrontBoardServices                  0x00007fff36d4f8e6 -[FBSSerialQueue _performNextFromRunLoopSource] + 22
	50  CoreFoundation                      0x00007fff23da1c91 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	51  CoreFoundation                      0x00007fff23da1bbc __CFRunLoopDoSource0 + 76
	52  CoreFoundation                      0x00007fff23da1394 __CFRunLoopDoSources0 + 180
	53  CoreFoundation                      0x00007fff23d9bf8e __CFRunLoopRun + 974
	54  CoreFoundation                      0x00007fff23d9b8a4 CFRunLoopRunSpecific + 404
	55  GraphicsServices                    0x00007fff38c05bbe GSEventRunModal + 139
	56  UIKitCore                           0x00007fff49372964 UIApplicationMain + 1605
	57  Cube                                0x000000010c5ca556 main + 54
	58  ???                                 0x00000001114244fe 0x0 + 4584523006
)
```

The issue can be reproduced by trying to run the cube demo on an iOS simulator with iOS13.7.

The current minimum supported iOS version  is iOS9 (as stated by Readme.md). What is the policy concerning this, should MoltenVK's minimum supported version be bumped up to iOS14 (in which case i'll have to stick to an older MoltenVK release) or should MoltenVK check for this selector's availability before calling ?

Thanks for the help.

[EDIT]
I am currently using MoltenVK-1.1.6

**Comment by billhollings** [comment]
The problems seems to be the logic around running an older version of iOS on the Simulator. Looks to be a simple fix.

**Comment by billhollings** [comment]
Since the device Simulator runs on the native host GPU, it can behave differently than when running on a device. Looking at the code, it seems unlikely that this problem would have been encountered on an iOS device under iOS 13 or earlier.

Nevertheless, PR #1520 includes a fix for both iOS and the device Simulator. Please verify and close this issue if it has fixed the problem you encountered.

**Comment by flolag** [comment]
Tested this morning on simulator, all good. Still needs to be tested on device, then I'll close the issue

**Comment by flolag** [comment]
Sorry I spoke too fast, while the fix works on iOS13 simulator, it doesn't on iOS14 & iOS15. It now gets stuck during frame presentation. It hangs within `MVKQueuePresentSurfaceSubmission::execute` on line 540 of `MVKQueue.mm` waiting on a semaphore.

I tested with the code from the v1.7 tag.

I don't think there's anything wrong in my synchronization setup (since it works on iOS13). Any clue ?

**Comment by billhollings** [comment]
> Sorry I spoke too fast, while the fix works on iOS13 simulator, it doesn't on iOS14 & iOS15. It now gets stuck during frame presentation. It hangs within MVKQueuePresentSurfaceSubmission::execute on line 540 of MVKQueue.mm waiting on a semaphore.

It's hard to see how this fix would trigger this. Do you have a call stack trace, or sample app we can test that provides more info?

**Comment by flolag** [comment]
Hi,

Unfortunately our project is quite complex, and sending you hundreds of thousands line of code wouldn't help. I'll try to provide more information once I get time to do it. Meanwhile I'll have to revert to using a patched version of tag 1.1.6

I'm marking this issue as closed since the issue at hand (the call to `-[MTLSimDevice counterSets]`) does not occur any more.

Thanks for your help.
